specifyweb/stored_queries/format.py

```diff
@@ -113,20 +113,34 @@ def objformat(self, query: QueryConstruct, orm_table: SQLTable,
         logger.debug("using dataobjformatter: %s",
                      ElementTree.tostring(formatterNode))
 
+        cycle_with_self = [*cycle_detector, (inspect(orm_table).class_.__name__, 'formatting')]
+
         def make_expr(query: QueryConstruct, fieldNodeText, fieldNodeAttrib) -> Tuple[
             QueryConstruct, blank_nulls, QueryFieldSpec]:
             path = fieldNodeText.split('.')
             path = [inspect(orm_table).class_.__name__, *path]
             formatter_field_spec = QueryFieldSpec.from_path(path)
             if formatter_field_spec.is_relationship():
+                next_table_name = formatter_field_spec.table.name
```

Comment:
this looks like the same algorithm front-end uses for cycle detection 👍:

https://github.com/specify/specify7/blob/a4a66e63a8ec27c82c60bca0c01bc0e1410b0725/specifyweb/frontend/js_src/lib/components/Formatters/formatters.ts#L147-L154C9

Reply:
Yeah, except that you can detect cycles more definitely on the frontend rather than the backend. This is because frontend can check if the id is repeated. Backend cannot since the query hasn't been run yet.